FWCore/ParameterSet/test/test_catch2_ps.cc

```diff
@@ -0,0 +1,875 @@
+    test_for_name<std::vector<int>>();
+
+    test_for_name<unsigned int>();
+    test_for_name<std::vector<unsigned int>>();
+
+    test_for_name<double>();
+    test_for_name<std::vector<double>>();
+
+    test_for_name<std::string>();
+    test_for_name<std::vector<std::string>>();
+    test_for_name<edm::ParameterSet>();
+    test_for_name<std::vector<edm::ParameterSet>>();
+
+    // Can't make default FileInPath objects...
+
+    // Now make sure that if we put in a parameter of type A, we don't
+    // see it when we ask for names of type B != A.
+    {
+      edm::ParameterSet p;
+      p.addParameter<double>("a", 2.5);
+      p.registerIt();
+      const bool tracked = true;
+      std::vector<std::string> names = p.getParameterNamesForType<int>(tracked);
+      REQUIRE(names.empty());
+    }
+  }
+
+  SECTION("Embedded PSet") {
+    edm::ParameterSet ps;
+    edm::ParameterSet psEmbedded, psDeeper;
+    psEmbedded.addUntrackedParameter<std::string>("p1", "wham");
+    psEmbedded.addParameter<std::string>("p2", "bam");
+    psDeeper.addParameter<int>("deepest", 6);
+    psDeeper.registerIt();
+    edm::InputTag it("label", "instance");
+    std::vector<edm::InputTag> vit;
+    vit.push_back(it);
+    psEmbedded.addParameter<edm::InputTag>("it", it);
+    psEmbedded.addParameter<std::vector<edm::InputTag>>("vit", vit);
+    psEmbedded.addParameter<edm::ParameterSet>("psDeeper", psDeeper);
+    psEmbedded.registerIt();
+    ps.addParameter<edm::ParameterSet>("psEmbedded", psEmbedded);
+    ps.addParameter<double>("topLevel", 1.);
+    ps.addUntrackedParameter<unsigned long long>("u64", 64);
+
+    std::vector<edm::ParameterSet> vpset;
+    edm::ParameterSet pset1;
+    pset1.addParameter<int>("int1", 1);
+    edm::ParameterSet pset2;
+    pset2.addParameter<int>("int2", 2);
+    edm::ParameterSet pset3;
+    pset3.addParameter<int>("int3", 3);
+    vpset.push_back(pset1);
+    vpset.push_back(pset2);
+    vpset.push_back(pset3);
+    ps.addParameter<std::vector<edm::ParameterSet>>("psVPset", vpset);
+
+    ps.registerIt();
+
+    std::string rep = ps.toString();
+    edm::ParameterSet defrosted(rep);
+    defrosted.registerIt();
+    edm::ParameterSet trackedPart(ps.trackedPart());
+
+    REQUIRE(defrosted == ps);
+    REQUIRE(trackedPart.exists("psEmbedded"));
+    REQUIRE(trackedPart.getParameterSet("psEmbedded").exists("p2"));
+    REQUIRE(!trackedPart.getParameterSet("psEmbedded").exists("p1"));
+    REQUIRE(trackedPart.getParameterSet("psEmbedded").getParameterSet("psDeeper").getParameter<int>("deepest") == 6);
+    REQUIRE(ps.getUntrackedParameter<unsigned long long>("u64") == 64);
+    REQUIRE(!trackedPart.exists("u64"));
+    std::vector<edm::ParameterSet> const& vpset1 = trackedPart.getParameterSetVector("psVPset");
+    REQUIRE(vpset1[0].getParameter<int>("int1") == 1);
+    REQUIRE(vpset1[1].getParameter<int>("int2") == 2);
+    REQUIRE(vpset1[2].getParameter<int>("int3") == 3);
+    SECTION("deprecated pset encoding") {
```

Comment:
```suggestion
    REQUIRE(vpset1[2].getParameter<int>("int3") == 3);

    SECTION("deprecated pset encoding") {
```
to make it more clear that a new `SECTION` begins.